File: Kotlin/src/collatz.kt
fun main() {
    val inputNumber = 27
    println(collatzIteration(inputNumber))
    collatzRecursive(inputNumber)
}

fun collatzIteration(number: Int): Int {
    var steps = 0
    var tmp = number
    while (tmp != 1) {
        print("$tmp ")
        tmp = if (tmp % 2 == 0) {
            tmp / 2
        } else {
            3 * tmp + 1
        }
        steps++
    }
    println(tmp)
    return steps
}

fun collatzRecursive(number: Int) {
    print("$number ")
    if (number == 1) {
        return
    }
    if (number % 2 == 0) {
        collatzRecursive(number / 2)
    } else {
        collatzRecursive(3 * number + 1)
    }
}

File: Rust/src/collatz.rs
fn main() {
    let input_number: i32 = 27;
    println!("{}", collatz_iteration(input_number));
    collatz_recursive(input_number);
}

fn collatz_iteration(number: i32) -> i32 {
    let mut steps: i32 = 0;
    let mut tmp: i32 = number;
    while tmp != 1 {
        print!("{} ", tmp);
        if tmp % 2 == 0 {
            tmp = tmp / 2;
        } else {
            tmp = 3 * tmp + 1;
        }
        steps = steps + 1;
    }
    println!("{}", tmp);
    return steps;
}

fn collatz_recursive(number: i32) {
    print!("{} ", number);
    if number == 1 {
        return;
    }
    if number % 2 == 0 {
        collatz_recursive(number / 2);
    } else {
        collatz_recursive(3 * number + 1);
    }
}
